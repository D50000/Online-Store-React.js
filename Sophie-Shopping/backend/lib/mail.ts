import { createTransport, getTestMessageUrl } from "nodemailer";

const transporter = createTransport({
  host: "smtp.ethereal.email",
  port: 587,
  auth: {
    user: "donny.beatty34@ethereal.email",
    pass: "shxZASXrt7NC2paQTv",
  },
});

function createSimpleEmail(text: string): string {
  return `
    <div className="email" style="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
    ">
      <h2>Hello There!</h2>
      <p>${text}</p>
      <p>Hi, Sophie Shopping's User !!! </p>
    </div>
  `;
}

export interface MailResponse {
  accepted?: string[] | null;
  rejected?: null[] | null;
  envelopeTime: number;
  messageTime: number;
  messageSize: number;
  response: string;
  envelope: Envelope;
  messageId: string;
}

export interface Envelope {
  from: string;
  to?: string[] | null;
}

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  const info = (await transporter.sendMail({
    to,
    from: "test@test.com",
    subject: "Your password reset token!",
    html: createSimpleEmail(`Your Password Reset Token Is Here!
    <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click Here to reset</a>
  `),
  })) as MailResponse;

  if (process.env.MAIL_USER.includes("ethereal.email")) {
    console.log(`ðŸ’Œ Message Sent!  Preview it at ${getTestMessageUrl(info)}`);
  }
}
