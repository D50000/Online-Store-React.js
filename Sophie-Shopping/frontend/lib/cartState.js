import { createContext, useContext, useState } from 'react';
import PropTypes from 'prop-types';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // Custom provider can access it via the consumer.

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// Make the custom hook for these functions.
function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };

CartStateProvider.propTypes = {
  children: PropTypes.any,
};
