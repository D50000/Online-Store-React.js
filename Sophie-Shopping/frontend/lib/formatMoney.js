export default function formatMoney(amount = 0) {
  const options = {
    style: 'currency',
    currency: 'TWD',
    minimumFractionDigits: 0,
  };

  const formatter = Intl.NumberFormat('zh-US', options);

  return formatter.format(amount);
}
