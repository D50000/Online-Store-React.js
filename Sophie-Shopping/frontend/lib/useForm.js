import { useState } from 'react';

export default function useForm(initial = {}) {
  const [inputs, setInputs] = useState(initial);

  function handleChange(e) {
    // Handle the data type that pass in.
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      value[0] = e.target.files;
    }
    setInputs({
      ...inputs,
      // Dynamic update the key and value
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const emptyState = Object.fromEntries(
      Object.entries(inputs).map(([key]) => [key, ''])
    );
    setInputs(emptyState);
  }

  // Return the initialState and SetStateAction()
  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
