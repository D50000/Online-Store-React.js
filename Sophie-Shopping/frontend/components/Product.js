import Link from 'next/link';
import PropTypes from 'prop-types';

import formatMoney from '../lib/formatMoney';

import ItemStyles from './styles/ItemStyles';
import Title from './styles/Title';
import PriceTag from './styles/PriceTag';

export default function Product({ product }) {
  return (
    <ItemStyles>
      <img
        src={product?.photo?.image?.publicUrlTransformed}
        alt={product.name}
      />
      <Title>
        <Link href={`/product/${product.id}`}>{product.name}</Link>
      </Title>
      <PriceTag>{formatMoney(product.price)}</PriceTag>
      <p>{product.description}</p>
      {/* TODO: Add btns to edit/delete */}
      <div className="buttonList">
        <Link
          // next.js routing syntax
          href={{
            pathname: 'update',
            query: {
              id: product.id,
            },
          }}
        >
          Edit !
        </Link>
      </div>
    </ItemStyles>
  );
}

Product.propTypes = {
  product: PropTypes.any,
};
