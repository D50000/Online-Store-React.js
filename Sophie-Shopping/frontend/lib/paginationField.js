import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // Setup Apollo keys myself.
    read(existing = [], { args, cache }) {
      // console.log({ existing, args, cache });
      const { skip, first } = args;

      // Read the detail of items on the page from the cache.
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // Check if we have existing items.
      const items = existing.slice(skip, skip + first).filter((item) => item);

      // On the last page.
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      // No item, fetch it from the DB.
      if (items.length !== first) {
        return false;
      }
      if (items.length) {
        return items;
      }

      // Fallback to network
      return false;
    },
    // Apollo fetch back the data will run merge()
    merge(existing, incoming, { args }) {
      const { skip, first } = args;

      // Calculate the item need to update.
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; i + 1) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}
