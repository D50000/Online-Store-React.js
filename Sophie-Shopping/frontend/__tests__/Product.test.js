import React from 'react';
import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';

import Product from '../components/Product';
import { fakeItem } from '../lib/testUtils';
import { CartStateProvider } from '../lib/cartState';

const product = fakeItem(); // Initial dta
describe('<Product/>', () => {
  it('renders out the price tag and title', () => {
    // Act
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider>
          <Product product={product} />
        </MockedProvider>
      </CartStateProvider>
    );
    // Assert
    debug(); // Debug will convert the React Node to html structure.
    const priceTag = screen.getByText('NT$5,000'); // 'screen' is the result of the render.
    // Expect
    expect(priceTag).toBeInTheDocument();

    const link = container.querySelector('a');
    debug(link);
    expect(link).toHaveAttribute('href', '/product/abc123');
    expect(link).toHaveTextContent(product.name);
  });
});
