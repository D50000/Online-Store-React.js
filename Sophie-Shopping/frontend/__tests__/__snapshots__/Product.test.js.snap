// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> Renders and matches the snapshot 1`] = `
<div>
  <div
    class="ItemStyles__Item-sc-hnkko-0 bLvtez"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <h3
      class="Title-sc-1x3vh0k-0 gyCUcc"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="PriceTag-sc-ji8zd7-0 jdqrss"
    >
      NT$5,000
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=abc123"
      >
        Edit !
      </a>
      <button
        type="button"
      >
        Add
         To Cart 
        -&gt;
      </button>
      <button
        type="button"
      >
        Delete !!
      </button>
    </div>
  </div>
</div>
`;
